### ISSUE TITLE
Self hosting on podman

### ISSUE THREAD

**Issue by afruizc** [plausible/analytics#71]
First of all, thank you for the work on plausible and especially the recent effort on self-hosting.

I am using podman instead of docker to run the containers. When running the main container, I get the following error: 

```
# podman run --name plausible-eph --hostname plausible --pod plausible plausible:http db migrate
Loading plausible..
Starting dependencies..
Starting repos..
Running migrations for plausible
** (Postgrex.Error) ERROR 42704 (undefined_object) type "regnamespace" does not exist
    (ecto_sql 3.4.4) lib/ecto/adapters/sql.ex:593: Ecto.Adapters.SQL.raise_sql_call_error/1
    (elixir 1.10.3) lib/enum.ex:1396: Enum."-map/2-lists^map/1-0-"/2
    (ecto_sql 3.4.4) lib/ecto/adapters/sql.ex:686: Ecto.Adapters.SQL.execute_ddl/4
    (ecto_sql 3.4.4) lib/ecto/migration/runner.ex:343: Ecto.Migration.Runner.log_and_execute_ddl/3
    (ecto_sql 3.4.4) lib/ecto/migration/runner.ex:117: anonymous fn/6 in Ecto.Migration.Runner.flush/0
    (elixir 1.10.3) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto_sql 3.4.4) lib/ecto/migration/runner.ex:116: Ecto.Migration.Runner.flush/0
    (stdlib 3.12.1) timer.erl:166: :timer.tc/1
```

I am using [this bash script](https://gist.github.com/afruizc/8fb076cbf58f853dc58594a6a5f0cd27) to start all the containers. I always start the main container manually to make sure the db and clickhouse are up.

Please let me know if I need to add more detail.

**Comment by ukutaht** [comment]
Hey, thanks for reporting.

Got my hands full trying to get the Docker version running in production at the moment. Will take a look at podman (had never heard of it before) once everything's working with standard Docker.

**Comment by tckb** [comment]
@afruizc that doesn't look like _plausible_ specific issue. Something related to postgresql, can you try it with latest version of postgresql? 

PS: I have never used podman myself but looked interesting :)

**Comment by makitsune** [comment]
Using the image `postgres:12.3-alpine` works perfectly for me!

**Comment by tckb** [comment]
Great! @afruizc what is the finalized config? May be you can add here the complete working config here ? It could be beneficial to others

**Comment by kolaente** [comment]
I got it working with this modified `docker-compose.yml`:

```yaml
version: "3.3"
services:
  db:
    image: postgres:12
    command: ["postgres", "-c", "log_statement=all", "-c", "log_destination=stderr"]
    volumes:
      - /apps/plausible/db:/var/lib/postgresql/data
    environment:
      - POSTGRES_PASSWORD=postgres
      - POSTGRES_DB=plausible
      - POSTGRES_USER=postgres
    restart: unless-stopped
  events-db:
    image: yandex/clickhouse-server
    restart: unless-stopped
  plausible:
    image: plausible/analytics
    command: sh -c "sleep 10 && /entrypoint.sh db createdb && /entrypoint.sh db migrate && /entrypoint.sh run"
    environment:
      - ENVIRONMENT=production
      - PORT=8080
      - SECRET_KEY_BASE=iYb1mP5cnmY+gUxo7C/h6XMigossPhzwd8/ic6LFnQ9Y58Fl1xduSWaPq0fHDdbn
      - SIGNING_SALT=PL/THF0VMOzuv1bOcldjDzYFBLryvXNs
      - HOST=localhost
      - DATABASE_URL=postgres://postgres:postgres@db:5432/plausible
      - DATABASE_TLS_ENABLED=false
      - ADMIN_USER_NAME=admin
      - ADMIN_USER_EMAIL=admin@plausible.local
      - ADMIN_USER_PWD=admin@1234!
      - APP_VERSION=test
      - MAILER_ADAPTER=Bamboo.SMTPAdapter
      - SMTP_HOST_ADDR=mail.example.com
      - SMTP_HOST_PORT=465
      - SMTP_USER_NAME=plausible@example.com
      - SMTP_USER_PWD=<password>
      - SMTP_HOST_SSL_ENABLED=true
      - SMTP_MX_LOOKUPS_ENABLED=false
      - CLICKHOUSE_DATABASE_HOST=events-db
      - CLICKHOUSE_DATABASE_NAME=plausible_events_db
    depends_on:
      - events-db
      - db
    restart: unless-stopped
    networks:
      - default
      - web
    labels:
      - "traefik.docker.network=web"
      - "traefik.enable=true"
      - "traefik.frontend.rule=Host:analytics.example.com"
      - "traefik.port=8080"
      - "traefik.protocol=http"

networks:
  web:
    external: true
```

However, it took quite some time to get it running since it would try the db migration, fail, retry, fail, etc. I let it do that for some time and eventually it succeeded.

**Comment by tckb** [comment]
@kolaente I am aware of this issue. Because during the first run the migration is already done and the data is already available. You need to delete the volume to run it again.  I mentioned this in the document: https://github.com/plausible/analytics/blob/master/HOSTING.md#up-and-running

Since, this was the first iteration, it is not perfect, sorry for the trouble ;) I have a working PR https://github.com/plausible/analytics/pull/89 I will address some of the issues there

**Comment by kolaente** [comment]
> Since, this was the first iteration, it is not perfect, sorry for the trouble ;) I have a working PR #89 I will address some of the issues there

No worries :)

I really like the project so far and am glad there is a self hosted version at all.

**Comment by makitsune** [comment]
I wanted to share my Docker compose file with Traefik which I've been happily using for the last couple of weeks.

```yml
# https://github.com/plausible/analytics/blob/master/HOSTING.md

version: "3.3"
networks:
    traefik:
        external:
            name: traefik
services:
    plausible_db:
        container_name: plausible_db
        restart: always
        image: postgres:12.3-alpine
        command:
            [
                "postgres",
                "-c",
                "log_statement=all",
                "-c",
                "log_destination=stderr",
            ]
        volumes:
            - ./db:/var/lib/postgresql/data
        environment:
            - POSTGRES_PASSWORD=postgres
            - POSTGRES_DB=plausible_db
            - POSTGRES_USER=postgres

    plausible_events_db:
        container_name: plausible_events_db
        restart: always
        image: yandex/clickhouse-server

    plausible:
        container_name: plausible
        restart: always
        image: plausible/analytics:latest
        command: sh -c "sleep 10 && /entrypoint.sh db createdb && /entrypoint.sh db migrate &&/entrypoint.sh run"
        environment:
            - HOST=plausible.example.com
            - SCHEME=https
            - PORT=8080
            - SECRET_KEY_BASE=...
            - SIGNING_SALT=...
            - ENVIRONMENT=production
            - APP_VERSION=latest

            # - ADMIN_USER_NAME=...
            # - ADMIN_USER_EMAIL=...
            # - ADMIN_USER_PWD=...

            - MAILER_ADAPTER=Bamboo.SMTPAdapter
            - SMTP_HOST_ADDR=smtp.gmail.com
            - SMTP_HOST_PORT=465
            - SMTP_USER_NAME=...
            - SMTP_USER_PWD=...
            - SMTP_HOST_SSL_ENABLED=true
            - SMTP_MX_LOOKUPS_ENABLED=false

            - DATABASE_URL=postgres://postgres:postgres@plausible_db:5432/plausible
            - DATABASE_TLS_ENABLED=false

            - CLICKHOUSE_DATABASE_HOST=plausible_events_db
            - CLICKHOUSE_DATABASE_NAME=plausible_events_db

            - GEOLITE2_COUNTRY_DB=/GeoLite2-Country.mmdb
        depends_on:
            - plausible_events_db
            - plausible_db
        links:
            - plausible_db
            - plausible_events_db
        volumes:
            - ./GeoLite2-Country.mmdb:/GeoLite2-Country.mmdb:ro
        labels:
            - traefik.enable=true
            - traefik.http.routers.plausible.rule=Host("plausible.example.com")
            - traefik.http.routers.plausible.entrypoints=websecure
            - traefik.http.routers.plausible.service=plausible
            - traefik.http.routers.plausible.middlewares=plausible-no-home,plausible-no-register
            - traefik.http.routers.plausible.tls.certresolver=le

            - traefik.http.services.plausible.loadbalancer.server.port=8080

            - traefik.http.middlewares.plausible-no-home.redirectregex.regex=^https:\/\/([^\/]+)\/?$$
            - traefik.http.middlewares.plausible-no-home.redirectregex.replacement=https://$$1/login

            - traefik.http.middlewares.plausible-no-register.redirectregex.regex=^https:\/\/([^\/]+)\/register\/?$$
            - traefik.http.middlewares.plausible-no-register.redirectregex.replacement=https://$$1/login

            - traefik.docker.network=traefik
        networks:
            - default
            - traefik
```

It redirects `/` and `/register` to `/login` so that no-one else can make an account.

Thank you so much for making plausible, it's really lovely and I hope I can help make it better and/or support it. It works well with Angular (uses `history.pushState`) which is a big relief.
 
The only part that confuses me as of now is if you set the admin user details as environment variables, it will fail to start if that user already exists (e.g. second run). When you don't specify the admin user, it will generate a new admin user every time. I probably have lots of unused admin accounts and only one that I actual use.

**Comment by tckb** [comment]
Hi @makitsune thanks am just a contributor like you :) In practice, you need to run the migration only once before      You start plausible. The idea is that you “prime” your DB and then you start. I did a poor job explaining that in the docs and the dockerfile was a sample to explain how everything connected


I will fix this issue in the next iteration pr#88 will fix these issues

**Comment by kolaente** [comment]
So it seems like clickhouse container looses all data upon container restart.

To fix this, it needs a volume which maps to `/var/lib/clickhouse` in the container like so:

```yaml
  events-db:
    image: yandex/clickhouse-server
    restart: unless-stopped
    volumes:
      - /mnt/gluster/apps/plausible/events-db:/var/lib/clickhouse
```

I'll send a PR.

**Comment by kolaente** [comment]
PR is up: #94

**Comment by tckb** [comment]
@ukutaht  / @metmarkosaric  wouldn't it be beter issues related to selfhosting are labelled together ? -- it will help in fixing all the issues together in PR https://github.com/plausible/analytics/pull/89

**Comment by metmarkosaric** [comment]
makes sense. I'll do that!

**Comment by tckb** [comment]
@makitsune I have test fixes for migration, you can test it at https://github.com/tckb/plausible/blob/master/HOSTING.md#up-and-running let me know if this fixes your issue

**Comment by mannp** [comment]
Hi @makitsune thanks for the home and register redirects for traefik.... is it for traefik 1.x or 2.x as I am getting a redirect to https://$1/login with 2.x, so it doesn't appear to be picking up the host variable in your redirects?

**Comment by tckb** [comment]
I am not sure if this one will be addressed due to the limitations of `podman`  may be we can check it once it has better support for all platforms. should we close this one? @metmarkosaric

**Comment by metmarkosaric** [comment]
Sounds good @tckb, thanks!
